core/src/io/anuke/mindustry/game/PlayerLog.java

```diff
@@ -0,0 +1,254 @@
+package io.anuke.mindustry.game;
+
+import io.anuke.arc.collection.ObjectSet;
+import io.anuke.arc.math.geom.Rectangle;
+import io.anuke.arc.util.Time;
+import io.anuke.mindustry.entities.type.Player;
+import io.anuke.mindustry.world.Tile;
+
+import java.util.Arrays;
+
+/**
+ * Maintains a fixed-window log of player actions, primarily to aid griefer detection.
+ */
+public class PlayerLog{
+    public enum Action{
+        constructed,
+        deconstructed,
+        linked,
+        unlinked,
+        configured;
+
+        public static final String LIST = list();
+
+        public static String list(){
+            final StringBuilder sb = new StringBuilder();
+            for(Action value : values()){
+                sb.append(", ").append(value);
+            }
+            return sb.toString().substring(2);
+        }
+    }
+    /** The size of the log; for 16 players averaging 200 APM this will save about 5 min of actions */
+    public static final int BUFFER_SIZE = 16*1024;
```

Comment:
This should be in `camelCase`. I think I might have mentioned it earlier. Yes, it's not standard capitalization, but I believe making harmless constant all caps is ugly, and it doesn't match the rest of the codebase.